Change high frequency free running timers to on-demand tasks

omerfirmak opened the pull request:
We could maybealso increase their firing delay from 10 msecs to something more reasonable

omerfirmak committed:
Client: Enable gossip timer on-demand

A free running 10 msec timer per client causes WAY to much context
switches. That's just wasted CPU cycles.

omerfirmak committed:
Nominator: Enable envelope timer on-demand

a free running 10 msecs timer would cause too much unnecessary
context switches.

Geod24 commented:
I think you can reuse the same instance, and call `rearm` on it, instead of calling `setTimer` again.

Geod24 commented:
Also if it's not period, do you need to `stop` it ?

omerfirmak commented:
> 
> 
> I think you can reuse the same instance, and call `rearm` on it, instead of calling `setTimer` again.

Our common interface only has stop, that might be a vibe.d thing maybe.

https://github.com/bosagora/agora/blob/236b2ee0d98254c498bef5eda9ef5539fd785b10/source/agora/common/Task.d#L96-L105

omerfirmak commented:
> 
> 
> Also if it's not period, do you need to `stop` it ?

If it is enabled (non-null timer), we dont want it to fire one last time after `shutdown`. This might not work as expected with localrest tho, I'll check.

omerfirmak commented:
Ok, it works fine in LocalRest too

https://github.com/Geod24/localrest/blob/3b650c53ea8af30061a7feab5af43992ad88655b/source/geod24/LocalRest.d#L537-L538

Geod24 commented:
> Our common interface only has stop

Well, you know what to do :D

Geod24 commented:
> If it is enabled (non-null timer), we dont want it to fire one last time after `shutdown`.

I agree, but that would only happen if we re-enter the event loop.

omerfirmak commented:
> 
> 
> > If it is enabled (non-null timer), we dont want it to fire one last time after `shutdown`.
> 
> I agree, but that would only happen if we re-enter the event loop.

Then why call it for the periodic one either? I think it's better being safe than sorry :D

omerfirmak committed:
Add rearm() support to Vibe.d&LocalRest timers

omerfirmak committed:
Update localrest

Geod24 commented on source/agora/network/Client.d:
```diff
@@ -168,7 +171,9 @@ public class NetworkClient
             format("Request failure to %s after %s attempts", address,
                 max_retries));
-        this.gossip_timer = this.taskman.setTimer(10.msecs, &this.gossipTask, Periodic.Yes);
+        // Create and stop timer immediately
+        this.gossip_timer = this.taskman.setTimer(GossipDelay, &this.gossipTask, Periodic.No);
+        this.gossip_timer.stop();
```
https://github.com/vibe-d/vibe-core/blob/9ca874dc75020b1488a0ac64a1cc5f6471ba15f4/source/vibe/core/core.d#L1243-L1255

But okay, this works :D

Geod24 commented on source/agora/consensus/protocol/Nominator.d:
```diff
@@ -189,12 +192,14 @@ extern(D):
         this.enroll_man = enroll_man;
         this.scp_envelope_store = new SCPEnvelopeStore(cacheDB);
+        // Create and stop timer immediately
+        this.envelope_timer = this.taskman.setTimer(10.msecs,
```
`EnvTaskDelay` ?

Geod24 reviewed (approved):
LGTM. The `rearm` thing is dangerous, because adding a new method can go wrong very easily, but luckily this doesn't change much.

Geod24 commented:
Did the minor change directly, merging

omerfirmak committed:
Flash: Set quorum threshold to %100

these tests fail often because network decides
to externalize an empty block from time to time.

Setting the quorum threshold to %100 will hopefully
help.
